File: c++/usaco/bronze/socdist1.cpp
#include <bits/stdc++.h>
#define ll long long
using namespace std;

void setIO(string s) {
  ios_base::sync_with_stdio(0);
  cin.tie(0);
  freopen((s + ".in").c_str(), "r", stdin);
  freopen((s + ".out").c_str(), "w", stdout);
}

ll largestInteriorGap(string s, ll &gapStart) {
  ll biggestGap = 0, currentStart = -1, n = s.length();
  for (ll i = 0; i < n; i++) {
    if (s[i] == '1') {
      if (currentStart != -1 && i - currentStart > biggestGap) {
        biggestGap = i - currentStart;
        gapStart = currentStart;
      }
      currentStart = i;
    }
  }
  return biggestGap;
}

ll smallestInteriorGap(string s) {
  ll smallestGap = LONG_LONG_MAX, currentStart = -1, n = s.length();
  for (ll i = 0; i < n; i++) {
    if (s[i] == '1') {
      if (currentStart != -1 && i - currentStart < smallestGap) {
        smallestGap = i - currentStart;
      }
      currentStart = i;
    }
  }
  return smallestGap;
}

ll tryCowInLargestGap(string s) {
  ll gapStart, largestGap = largestInteriorGap(s, gapStart);
  if (largestGap >= 2) {
    s[gapStart + largestGap / 2] = '1';
    return smallestInteriorGap(s);
  }
  return -1;
}

int main() {
  setIO("socdist1");
  ll n;
  string s, temp;
  ll answer = 0;
  cin >> n >> s;
  ll gapStart, largestGap = largestInteriorGap(s, gapStart);

  // 1. two cows in one gap
  if (largestGap >= 3) {
    temp = s;
    temp[gapStart + largestGap / 3] = '1';
    temp[gapStart + largestGap * 2 / 3] = '1';
    answer = max(smallestInteriorGap(temp), answer);
  }

  // 2. cows on either ends
  if (s[0] == '0' && s[n - 1] == '0') {
    temp = s;
    temp[0] = temp[n - 1] = '1';
    answer = max(smallestInteriorGap(temp), answer);
  }

  // 3. cow in interior and one on the left
  if (s[0] == '0') {
    temp = s;
    temp[0] = '1';
    answer = max(answer, tryCowInLargestGap(temp));
  }

  // 4. cow in interior and one on the right
  if (s[n - 1] == '0') {
    temp = s;
    temp[n - 1] = '1';
    answer = max(answer, tryCowInLargestGap(temp));
  }

  // 5. cows both in interior but not in same gap
  if (largestGap >= 2) {
    temp = s;
    temp[gapStart + largestGap / 2] = '1';
    answer = max(answer, tryCowInLargestGap(temp));
  }

  cout << answer << "\n";
  return 0;
}

File: c++/usaco/bronze/socdist1.py
import sys

sys.stdin = open('input.txt', 'r')
# sys.stdin = open('socdist1.in', 'r')
# sys.stdout = open('socdist1.out', 'w')

n = int(input())
stalls = input()

def largest_gap(s):
  cur = -1
  gap = 0
  start = 0
  for i in range(n):
    if s[i] == '1':
      if cur != -1 and i - cur > gap:
        biggest_gap = i - cur
        start = cur
      cur = i
  return (gap, start)

def smallest_gap(s):
  cur = -1
  gap = float('inf')
  for i in range(n):
    if s[i] == '1':
      if cur != -1 and i - cur < gap:
        smallest_gap = i - cur
      cur = i
  return gap

def try_in_largest_gap(s):
  gap_start, gap = largest_gap(s)
  copy_stalls = stalls
  if gap >= 2:
    copy_stalls[start + gap // 2] = '1'
    return smallest_gap()
  return -1

start, gap = largest_gap(stalls)
ans = 0
if gap >= 3:
  temp = stalls
  temp[start + gap / 3] = '1'
  temp[start + gap * 2 / 3] = '1'
  ans = max(smallest_gap(temp), ans)

if stalls[0] == '0' and stalls[-1] == '0':
  temp = stalls
  temp[0] = temp[-1] = '1'
  ans = max(smallest_gap(temp), ans)

if s[0] == '0':
  temp = stalls
  temp[0] = '1'
  ans = max(ans, try_in_largest_gap(temp))

if s[-1] == '0':
  temp = stalls
  temp[-1] = '1'
  ans = max(ans, try_in_largest_gap(temp))

if largest_gap >= 2:
  temp = stalls
  temp[start + gap // 2] == '1'
  ans = max(ans, try_in_largest_gap(temp))

print(ans)